exp/runtime/internal/controllers/extension_controller.go

```diff
@@ -21,74 +21,169 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1beta1"
+	runtimeclient "sigs.k8s.io/cluster-api/internal/runtime/client"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
 )
 
 // +kubebuilder:rbac:groups=runtime.cluster.x-k8s.io,resources=extension,verbs=get;list;watch;create;update;patch;delete
 
-// ExtensionReconciler reconciles a Extension object.
-type ExtensionReconciler struct {
+// Reconciler reconciles an Extension object.
+type Reconciler struct {
 	Client        client.Client
 	RuntimeClient runtimeclient.Client
-	Registry      registry.Registry
+
+	// WatchFilterValue is the label value used to filter events prior to reconciliation.
+	WatchFilterValue string
 }
 
-func (r *ExtensionReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
 	err := ctrl.NewControllerManagedBy(mgr).
-		For(&expv1.Extension{}).
+		For(&runtimev1.Extension{}).
+		Named("runtime/extension").
 		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
 		Complete(r)
-
 	if err != nil {
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
 	return nil
 }
 
-func (r *ExtensionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	log := ctrl.LoggerFrom(ctx)
-	log.V(4).Info("Reconciling")
+	if !r.RuntimeClient.IsRegistryReady() {
+		err := r.syncRegistry(ctx)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "extensions controller not initialized")
+		}
+		// Note: there was an intermittent bug when allowing this to continue (likely due to patching the same object multiple times?)
+		return ctrl.Result{}, nil
+	}
 
-	ext := &expv1.Extension{}
-	err := r.Client.Get(ctx, req.NamespacedName, ext)
+	log.V(4).Info("Reconciling")
+	extension := &runtimev1.Extension{}
+	err := r.Client.Get(ctx, req.NamespacedName, extension)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+			// Object not found, return.
 			return ctrl.Result{}, nil
 		}
 
 		// Error reading the object - requeue the request.
 		return ctrl.Result{}, err
 	}
 
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(extension, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// FIXME: could dedupe these patch calls. Ignoring for simplicity at this stage.
```

Comment:
Q: which patch calls could we dedupe?

Reply:
I think I understood it now. The ones in list vs the current item?

Reply:
Yeah, exactly. Like I said here it's fine to ignore for now.

Reply:
Yup